src/utils/imsettings-switch.1
.\" -*- coding: us-ascii -*-
.if \n(.g .ds T< \\FC
.if \n(.g .ds T> \\F[\n[.fam]]
.de URL
\\$2 \(la\\$1\(ra\\$3
..
.if \n(.g .mso www.tmac
.TH imsettings-switch 1 "Apr 10, 2013" "" ""
.SH NAME
imsettings-switch \- Switch Input Method
.SH SYNOPSIS
'nh
.fi
.ad l
\fBimsettings-switch\fR \kx
.if (\nx>(\n(.l/2)) .nr x (\n(.l/5)
'in \n(.iu+\nxu
[
\fB-hnqrx\fR
] [
\fB--no-update\fR
] [
\fB--quiet\fR
] [
\fB--restart\fR
] [
\fB--read-xinputrc\fR
] [
\fB--help\fR
] [Input Method name | configuration filename]
'in \n(.iu-\nxu
.ad b
'hy
.SH DESCRIPTION
\fBimsettings-switch\fR switches Input Method which is managed under imsettings. Either \fIInput Method name\fR or \fIconfiguration filename\fR are required unless \*(T<\fB\-\-read\-xinputrc\fR\*(T> option is specified to determine the target from \*(T<\fI$HOME/.config/imsettings/xinputrc\fR\*(T> or \*(T<\fI$xinputrcdir/xinputrc\fR\*(T>.
.SH OPTIONS
This program follows the usual GNU command line syntax, with long options starting with two dashes (`-'). A summary of options is included below.
.TP 
\*(T<\fB\-n\fR\*(T> \*(T<\fB\-\-no\-update\fR\*(T> 
Do not update \*(T<\fI$HOME/imsettings/xinputrc\fR\*(T>. without this option, \fBimsettings-switch\fR will updates it with the targeted Input Method.
.TP 
\*(T<\fB\-q\fR\*(T> \*(T<\fB\-\-quiet\fR\*(T> 
Do not show any messages except errors.
.TP 
\*(T<\fB\-r\fR\*(T> \*(T<\fB\-\-restart\fR\*(T> 
Restart the given Input Method or one which is determined from xinputrc by \*(T<\fB\-\-read\-xinputrc\fR\*(T> option.
.TP 
\*(T<\fB\-x\fR\*(T> \*(T<\fB\-\-read\-xinputrc\fR\*(T> 
Determine the targeted Input Method from \*(T<\fI$HOME/imsettings/xinputrc\fR\*(T> or \*(T<\fI$xinputrcdir/xinputrc\fR\*(T> if not available.
.TP 
\*(T<\fB\-h\fR\*(T> \*(T<\fB\-\-help\fR\*(T> 
Show summary of options
.TP 
\fIInput Method name\fR 
Switch Input Method to \fIInput Method name\fR.
.TP 
\fIconfiguration filename\fR 
Switch Input Method to \fIconfiguration filename\fR.
.SH "SEE ALSO"
\fBimsettings-list\fR(1)
\fBimsettings-reload\fR(1)
.SH AUTHOR
This manual page was written by Akira TAGOH <\*(T<akira@tagoh.org\*(T>>


src/NEWS
1.6.2 -> 1.6.3
=================
* Bug Fix:
  * Include man pages into the archive.

1.6.1 -> 1.6.2
=================
* Bug Fixes:
  * Fix the desktop detection for gnome-classic
  * Shut down imsettings-daemon if current desktop isn't targeted
  * Fix the bootstrap issue with recent automake
* Enhancement:
  * Add man pages

1.6.0 -> 1.6.1
=================
* Bug Fixes:
  * check the exit status for imsettings-target-checker.sh
  * Fix gtk+ support detection for gnome-classic session
  * obsolete NOT_RUN and lookup gsettings key for GNOME to determine
    if imsettings should manage input methods.

1.5.1 -> 1.6.0
=================
* Bug Fixes:
  * imsettings-xim: Fix a crash issue at finalizing.
  * Fix erroneous code.
  * Fix deprecated warnings from GLib.
* Enhancement:
  * Add Cinnamon desktop support

1.5.0 -> 1.5.1
=================
* Bug Fixes:
  * Fix broken pkgconfig files.
  * Fix the build issue when srcdir != builddir.
  * Revert the previos change that run imsettings-switch on GNOME3 only when
    the session is gnome-fallback.
* Support libgxim 0.4.0

1.4.0 -> 1.5.0
=================
* Enhancement:
  * Add a MATE (>= 1.5) desktop support
* Improvement:
  * better fallback immodule support detection
* Bug Fixes:
  * move imsettings-check into libexecdir.
  * run imsettings-switch on GNOME3, only when the session is gnome-fallback
* Translation updates

1.3.1 -> 1.4.0
=================
* Enhancements:
  * Add a MATE desktop support
  * Update documents
* Translation updates

1.3.0 -> 1.3.1
=================
* Enhancement:
  * support GNOME3 in NOT_RUN variable.
* Translations updates

1.2.9 -> 1.3.0
=================
* Bug Fixes:
  * Issue#30: follow XDG base directory standard for log file
  * Change $HOME/.xinputrc to $XDG_CONFIG_HOME/imsettings/xinputrc
* Enhancement:
  * Support NOT_RUN variable in xinput file.

1.2.8.1 -> 1.2.9
===================
* Bug Fix:
  * xfce: fallback to the default configuration file on XDG dirs if any.
* Translation update

1.2.8 -> 1.2.8.1
====================
* Translations updates

1.2.7.1 -> 1.2.8
===================
* Bug Fixes:
  - Fix fails on distcheck.
  - Fix possibly accessing a corrupted memory.
* Improvement:
  - output more information to .imsettings.log for debugging.

1.2.7 -> 1.2.7.1
===================
* Bug Fixes:
  - Fix DSO link issue.
  - Fix a build fail.

1.2.6 -> 1.2.7
===================
* Bug Fixes:
  * gconf backend: fix build issue
  * library: memory leaks
* Enhancements:
  * Support multiple immodule specifier (gtk+ >= 3.3.3)
  * Add GObject-introspection support

1.2.5 -> 1.2.6
===================
* Bug Fix:
  * Do not unown the bus name on the shutdown.
* Enhancement:
  * Improve the desktop detection.

1.2.4 -> 1.2.5
===================
* Bug Fix:
  * Fix taking CPU load issue again.
* Translation updates.

1.2.3 -> 1.2.4
===================
* Bug Fixes:
  * Restart IM properly on reload.
  * Stop the daemon properly with SIGTERM.
  * Fix taking CPU load issue (rhbz#718092)

1.2.2 -> 1.2.3
===================
* Bug Fixes:
  * Fix a typo in the code to detect the gtk+ support.
  * Set some environment variables explicitly when bringing
    IM processes up. (rhbz#700513)
  * Fix the LXDE backend module to reload the configuration
    in LXDE properly on the latest desktop.
  * Fix a crash issue on killing the processes.
  * Fix a race condition issue on switching IM.
  * Fix an abort issue in logger. (rhbz#701431)
* Improvement:
  * Revamp Qt backend module.

1.2.1 -> 1.2.2
==================
* Bug Fixes:
  * Fix an issue keeping alive the defunct process.
  * Stop a notification when no backend modules available. (rhbz#693809)
* Improvements:
  * Better handling of detecting the supported toolkits.
  * show a warning on startup time if no backend modules
    is loaded in the daemon.
* Feature:
  * Add an option to check requirements of imsettings running separately.
* Translation Update:
  * Danish

1.2.0 -> 1.2.1
==================
* Bug Fixes:
  * stop imsettings-xim process when the module is unloaded.
  * invoke the proper IM at the startup when XDG autostart is enabled.
  * get rid of the unnecessary warnings on the command-line options.
  * fix a typo in xinput.sh.
* Translation Updates

1.1.0 -> 1.2.0
==================
* Feature:
  * small utility to self-check the working environment.
* Improvements:
  * Pop up the notice via libnotify if no modules are loaded.
  * Add imsettings documents generated by gtk-doc.
* Bug Fixes:
  * Get rid of the unnecessary warning during reading the config.
  * Give proper permissions to the scripts.
  * Call method to stop the service instead of the signal if it's supported.
  * Fix typos
  * Do not update .xinputrc if requested.
  * Improve a performance of imsettings-daemon. (rhbz#676813)

1.0.1 -> 1.1.0
==================
* Feature:
  * GNOME3 support.
* Improvements:
  * code clean up for compiler warnings.

1.0.0 -> 1.0.1
==================
* Docs updated.
* Enhancement:
  * Gentoo support (MATSUU Takuto)
* Bug Fixes:
  * Fix the handling of "User Specific".
  * Improve the response time.
  * Fix not-found issue on xim.conf when the upper/lower-case
    is randomly mixed up on the locale name.
  * Fix not enabling IM on Xfce 4.8.

0.108.1 -> 1.0.0
===================
* Simplify the process structures.
* Revise a lot of code.
* Modularize the backends.
* GDBus transition.
* Bump API version.
* Bug Fix:
  * rhbz#526205: Unable to find XIM with the certain locale

0.108.0 -> 0.108.1
=====================
* Bug Fixes:
  * rhbz#573604: Don't restart the IM process when the exit status is 0.
  * rhbz#582448: Fix failing to bring the process up.
  * rhbz#599924: Fix segfault caused by the race condition on popen.
  * rhbz#603582: Fix a segfault in imsettings-lxde-helper
* Translation updated:
  * Kannada
  * Korean
  * Russian
  * Traditional Chinese

0.107.4 -> 0.108.0
=====================
* Features:
  * Add LXDE support.
* Bug Fixes:

src/utils/imsettings-switch.c
/* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*- */
/* 
 * imsettings-switch.c
 * Copyright (C) 2008-2012 Red Hat, Inc. All rights reserved.
 * 
 * Authors:
 *   Akira TAGOH  <tagoh@redhat.com>
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth
 * Floor, Boston, MA  02110-1301  USA
 */
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <locale.h>
#include <stdlib.h>
#include <unistd.h>
#include <glib/gi18n.h>
#include "imsettings.h"
#include "imsettings-client.h"
#include "imsettings-utils.h"

int
main(int    argc,
     char **argv)
{
	IMSettingsClient *client = NULL;
	IMSettingsInfo *info = NULL;
	gchar *locale, *module = NULL;
	gboolean arg_force = FALSE, arg_no_update = FALSE, arg_restart = FALSE, arg_xinputrc = FALSE, arg_quiet = FALSE;
	GOptionContext *ctx = g_option_context_new(_("[Input Method name|xinput.conf]"));
	GOptionEntry entries[] = {
		/* For translators: this is a translation for the command-line option. */
		{"force", 'f', 0, G_OPTION_ARG_NONE, &arg_force, N_("Force restarting the IM process regardless of any errors."), NULL},
		/* For translators: this is a translation for the command-line option. */
		{"no-update", 'n', 0, G_OPTION_ARG_NONE, &arg_no_update, N_("Do not update the user xinputrc."), NULL},
		/* For translators: this is a translation for the command-line option. */
		{"quiet", 'q', 0, G_OPTION_ARG_NONE, &arg_quiet, N_("Shut up the extra messages."), NULL},
		/* For translators: this is a translation for the command-line option. */
		{"restart", 'r', 0, G_OPTION_ARG_NONE, &arg_restart, N_("Restart input method"), NULL},
		/* For translators: this is a translation for the command-line option. */
		{"read-xinputrc", 'x', 0, G_OPTION_ARG_NONE, &arg_xinputrc, N_("Read xinputrc to determine the input method"), NULL},
		{NULL, 0, 0, 0, NULL, NULL, NULL}
	};
	GError *error = NULL;
	int retval = 0, api_version;
	GVariant *v;
	GVariantIter *iter;
	const gchar *key, *val;

#ifdef ENABLE_NLS
	bindtextdomain (GETTEXT_PACKAGE, IMSETTINGS_LOCALEDIR);
#ifdef HAVE_BIND_TEXTDOMAIN_CODESET
	bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
#endif /* HAVE_BIND_TEXTDOMAIN_CODESET */
	textdomain (GETTEXT_PACKAGE);
#endif /* ENABLE_NLS */

	setlocale(LC_ALL, "");
	locale = setlocale(LC_CTYPE, NULL);

	g_type_init();

	g_option_context_add_main_entries(ctx, entries, GETTEXT_PACKAGE);
	if (!g_option_context_parse(ctx, &argc, &argv, &error)) {
		if (error != NULL) {
			g_printerr("%s\n", error->message);
		} else {
			g_warning(_("Unknown error in parsing the command lines."));
		}
		return 1;
	}

	if (!imsettings_is_enabled()) {
		g_printerr(_("IMSettings is disabled on the system.\n"));
		retval = 1;
		goto end;
	}
	client = imsettings_client_new(locale);
	if (!client) {
		g_printerr(_("Unable to create a client instance.\n"));
		retval = 1;
		goto end;
	}
	if ((api_version = imsettings_client_get_version(client, NULL, &error)) != IMSETTINGS_SETTINGS_API_VERSION) {
		if (error)
			goto error;
		g_printerr(_("Currently a different version of imsettings is running.\nRunning \"imsettings-reload\" may help but it will restart the Input Method\n"));
		retval = 1;
		goto end;
	}
	v = imsettings_client_get_module_settings(client, NULL, &error);
	if (error)
		goto error;
	g_variant_get(v, "a{ss}", &iter);
	if (!g_variant_iter_next(iter, "{&s&s}", &key, &val)) {
		g_warning(_("No backend modules available"));
		retval = 1;
		goto end;
	}
	g_variant_iter_free(iter);
	g_variant_unref(v);

	if (arg_restart) {
		const gchar *m, *l;

		info = imsettings_client_get_active_im_info(client, NULL, &error);
		if (error)
			goto error;
		m = imsettings_info_get_short_desc(info);
		l = imsettings_info_get_language(info);
		if (m == NULL || m[0] == 0 ||
		    strcmp(m, IMSETTINGS_NONE_CONF) == 0) {
			g_printerr(_("No Input Method running to be restarted.\n"));
			retval = 1;
			goto end;
		}
		imsettings_client_switch_im(client, IMSETTINGS_NONE_CONF, FALSE, NULL, &error);
		if (error)
			goto error;
		imsettings_client_set_locale(client, l);
		imsettings_client_switch_im(client, m, FALSE, NULL, &error);
		if (error)
			goto error;

		if (!arg_quiet)
			g_print(_("Restarted %s\n"), m);
		goto end;
	}
	if (argc < 2 && !arg_xinputrc) {
		gchar *help = g_option_context_get_help(ctx, FALSE, NULL);

		g_print("%s", help);
		g_free(help);
		goto end;
	}
	if (arg_xinputrc) {
		const gchar *env = g_getenv("IMSETTINGS_MODULE");

		if (env) {
			GVariant *v = imsettings_client_get_info_variant(client, env, NULL, NULL);

			if (v) {
				module = g_strdup(env);
				g_variant_unref(v);
			}
		}
		if (!module)
			module = imsettings_client_get_user_im(client, NULL, &error);
	} else {
		module = g_strdup(argv[1]);
	}

	imsettings_client_switch_im(client, module, !arg_no_update, NULL, &error);
	if (error) {
	  error:
		g_printerr("%s\n", error->message);
		if (g_error_matches(error,
				    IMSETTINGS_GERROR,
				    IMSETTINGS_GERROR_NOT_TARGETED_DESKTOP)) {
			g_clear_error(&error);
			imsettings_client_reload(client, api_version < 4, NULL, &error);
		}
		retval = 1;
		g_clear_error(&error);
		goto end;
	}
	if (!arg_quiet)
		g_print(_("Switched input method to %s\n"), module);
  end:
	g_free(module);
	if (info)
		g_object_unref(info);
	if (client)
		g_object_unref(client);
	g_option_context_free(ctx);

	return retval;
}
